### PR TITLE
Priority dropdown based on event

### PR DESCRIPTION
**Pull request by biplovbhandari**

### CODE REVIEW

**Comment by biplovbhandari** [comment]
@flavour Please merge this

File: controllers/cap.py [hunk 1]
```diff
@@ -416,6 +416,59 @@ def prep(r):
...
+                                                         limitby=(0, 1)).first()
+            try:
+                event_type_name = item.name
+            except:
+                result = current.xml.json_message(False, 400, "Event Type Not Found!")
+            else:
+                wptable = s3db.cap_warning_priority
+                query = (wptable.event_type == event_type_name)
+                      
+                rows = db(query).select(wptable.id, wptable.name,
```
Comment (flavour) [inline controllers/cap.py]: Would prefer each field on separate line, as per current code

File: controllers/cap.py [hunk 2]
```diff
@@ -416,6 +416,59 @@ def prep(r):
...
+                if rows:
+                    row_dict = [{"id": r.id, "name": T(r.name)} for r in rows] + \
+                               [{"id": "", "name": T("Undefined")}]
+                    result = jsons(row_dict)
+                else:
+                    row = db(wptable.event_type == "others").select(wptable.id, 
+                                                                    wptable.name,
+                                                                    orderby = wptable.id)
+            
+                    row_dict = [{"id": r.id, "name": T(r.name)} for r in row] + \
```
Comment (flavour) [inline controllers/cap.py]: Why iterate over a single row here? Surely more efficient to just build the row_dict directly?

File: modules/s3db/cap.py [hunk 3]
```diff
@@ -490,14 +491,55 @@ def model(self):
...
+                     Field("event_type",
+                           label = T("Event Type"),
+                           ),
+                     Field("urgency",
+                           requires = IS_IN_SET(cap_info_urgency_opts),
+                           ),
+                     Field("severity",
+                           requires = IS_IN_SET(cap_info_severity_opts),
+                           ),
+                     Field("certainty",
```
Comment (flavour) [inline modules/s3db/cap.py]: Why no label for these fields?
No label => no ability to localise

File: modules/s3db/cap.py [hunk 4]
```diff
@@ -537,7 +579,24 @@ def model(self):
...
                            ), # 1 or more allowed
-                     self.event_type_id(),
+                     self.event_type_id(requires = IS_ONE_OF(
+                                                    db, "event_event_type.id",
+                                                    S3Represent(
+                                                                lookup="event_event_type",
+                                                                translate=True
+                                                                ),
+                                                    orderby = "event_event_type.name",
+                                                    sort = True,
```
Comment (flavour) [inline modules/s3db/cap.py]: What is different about this requires to the default one? (I asked this before but got no answer)

**Comment by flavour** [comment]
Please ensure 1 commit per PR (rebase/compress)
Also a request to 'review' is nicer than the request to 'merge'

File: controllers/cap.py (hunk 2 above)
Comment (biplovbhandari) [inline controllers/cap.py]: @flavour this could be multiple rows...so I iterate

**Comment by biplovbhandari** [comment]
@flavour I have tried to make changes as per suggestions...Please review

File: controllers/cap.py (hunk 2 above)
Comment (flavour) [inline controllers/cap.py]: Ah, a bad variable name then! row => 1 & rows => multiple
So yes we need to iterate, but pls change the variable name accordingly

**Comment by biplovbhandari** [comment]
@flavour Please merge this

File: modules/s3db/cap.py [hunk 5]
```diff
@@ -490,14 +491,58 @@ def model(self):
             ("Unknown", T("Certainty unknown")),
         ])
 
+        # ---------------------------------------------------------------------
+        # Warning Priorities for CAP
+        
+        tablename = "cap_warning_priority"
+        define_table(tablename,
+                     Field("priority_rank", "integer",
+                           length = 2,
```
Comment (flavour) [inline modules/s3db/cap.py]: still missing a label on this field

File: modules/s3db/cap.py [hunk 6]
```diff
@@ -547,9 +600,13 @@ def model(self):
                            widget = S3MultiSelectWidget(),
                            ), # 0 or more allowed
                      Field("priority",
+                           represent = priority_represent,
                            requires = IS_EMPTY_OR(
-                                        IS_IN_SET(cap_info_priority_opts)
-                                        ),
+                                                  IS_ONE_OF(
+                                                            db, "cap_warning_priority.id",
```
Comment (flavour) [inline modules/s3db/cap.py]: The reason this was indented less before was to keep within the 80 char page width - I guess this exceeds if you indent to the full like this?

**Comment by biplovbhandari** [comment]
@flavour Please merge this

**Comment by biplovbhandari** [comment]
@flavour Please merge this

**Pull request merged**